### PR TITLE
crimson/os/seastore: fix denc device_spec_t, device_type_t and segment_type_t

### PR DESCRIPTION
**Pull request by cyx1231st**
* device_spec_t should not include denc and be packed at the same time
* the underlying type of enum:uint8_t should support denc_traits already

Signed-off-by: Yingxin Cheng <yingxin.cheng@intel.com>
Signed-off-by: luo rixin <luorixin@huawei.com>

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

File: src/crimson/os/seastore/device.h [hunk 1]
```diff
@@ -25,7 +25,7 @@ struct device_spec_t {
     denc(v.id, p);
     DENC_FINISH(p);
   }
-} __attribute__ ((packed));
+};
```
Comment (tchaikov) [inline src/crimson/os/seastore/device.h:28]: > device_spec_t should not include denc and be packed at the same time

if we only need one of them, why they are both dropped in this change?

Comment (cyx1231st) [reply src/crimson/os/seastore/device.h:28]: I think we should either encode and decode (`denc()`) `device_spec_t`, or pack it. Only `__attribute__ ((packed))` is droped for `device_spec_t`.

**Comment by cyx1231st** [comment]
jenkins test make check

File: src/crimson/os/seastore/device.h (hunk 1 above)
Comment (tchaikov) [reply src/crimson/os/seastore/device.h:28]: ah, sorry, i was wrong. should have read the commit message more carefully.

**Comment by cyx1231st** [comment]
jenkins test make check

**Comment by tchaikov** [comment]
jenkins test make check

**Pull request merged**